### PR TITLE
feature/products-device-add

### PR DESCRIPTION
**Pull request by busticated**
## Description

Adds `particle product device add` subcommand to add product devices ([ch](
https://app.clubhouse.io/particle/story/44994/enable-adding-new-product-devices))

## How to Test

_Note: requires you to have a Particle product with one or more devices_

Run `particle product device add --help` and explore the command and options.

## Completeness

- [x] User is totes amazing for contributing!
- [x] Contributor has signed [CLA](https://docs.google.com/a/particle.io/forms/d/1_2P-vRKGUFg5bmpcKLHO_qNZWGi5HKYnfrrkd-sbZoA/viewform)
- [x] Problem and solution clearly stated
- [x] Tests have been provided
- [x] Docs have been updated
- [x] CI is passing

### CODE REVIEW

File: src/cmd/product.js [hunk 1]
```diff
@@ -20,22 +23,86 @@ module.exports = class ProductCommand {
...
+			return this.addDevice({ params: { product, device } });
+		}
+
+		const msg = `Adding devices in ${file} to product ${product}`;
+		const upload = readDeviceListFile(file).then((identifiers) => uploadProductDevices(product, identifiers));
+		return this.showBusySpinnerUntilResolved(msg, upload)
+			.then(result => this.showDeviceAddResult(result));
+	}
+
+	showDeviceAddResult({ product, identifiers, status: { invalidDeviceIds = [], nonmemberDeviceIds = [] } } = {}){
```
Comment (wraithan) [inline src/cmd/product.js]: one of these days, seeing array/object defaults in JS wont make me panic like python taught me to

File: src/cmd/product.js [hunk 2]
```diff
@@ -20,22 +23,86 @@ module.exports = class ProductCommand {
...
+				'',
+				`Product ${product} Includes:`,
+				dedupeAndStringifyIDList(identifiers),
+				''
+			);
+		}
+
+		// TODO (mirande): it seems the API never returns this? currently
+		// adding a device id which is already in another product yields
+		// an api response flagging that id as `invalidDeviceIds`
```
Comment (wraithan) [inline src/cmd/product.js]: Looking at the test that is passing for API service around `nonmemberDeviceIds` it looks like that's returned when a device that is owned by another person who is not part of the org is provided as part of the set.

The filter that builds the `nonmemberDeviceIds` set confirms that:

```js
if (device.user_id || device.customer_id) {
	return device.belongsToOrg(org).then(function returnInvalid(belongs) {
		return !belongs;
	});
}
return false;
```

https://github.com/particle-iot/api-service/blob/master/src/contexts/productDeviceImportContext.js#L142

Comment (busticated) [reply src/cmd/product.js]: 🤔 i still can't get it to return anything in `nonmemberDeviceIds` - here's what i've tried:

my test user has a product, an org and an org product. when i try to add a device that belongs to me to either of my test user's products, i get an "invalid id" response. likewise, when i try to add one of my product devices to either those target products, i get the same error state. lastly, i tried adding a device which belongs to the capture test fleet (an org product) and got the same "invalid id" response.

cc @idokleinman any ideas?

File: test/e2e/product.e2e.js [hunk 3]
```diff
@@ -261,6 +264,160 @@ describe('Product Commands', () => {
...
+			'  particle product device add 12345 --file ./path/to/device_ids.txt  Adds a list of devices into product 12345'
+		];
+
+		before(async () => {
+			await cli.setTestProfileAndLogin();
+		});
+
+		beforeEach(async () => {
+			await Promise.all([
+				fs.writeFile(deviceIDsFilePath, [PRODUCT_01_DEVICE_01_ID, PRODUCT_01_DEVICE_02_ID].join(os.EOL), 'utf8'),
```
[the remark this answers is not in the document]
Comment (busticated) [reply test/e2e/product.e2e.js]: ✅

File: src/cmd/product.js [hunk 4]
```diff
@@ -83,3 +136,72 @@ function createAPI(){
...
+		.catch(error => {
+			if (error.code === 'ENOENT'){
+				throw new Error(`${file} does not exist`);
+			}
+			throw error;
+		});
+}
+
+function filterDeviceIdentifiers(identifiers, invalid = [], nonmember = []){
+	identifiers = Array.isArray(identifiers) ? identifiers : [identifiers];
```
[the remark this answers is not in the document]
Comment (busticated) [reply src/cmd/product.js]: ✅

File: src/cmd/product.js [hunk 5]
```diff
@@ -83,3 +136,72 @@ function createAPI(){
...
+			out.push(x);
+		}
+		return out;
+	}, []);
+}
+
+function hasDeviceIdentifier(x = '', ids = []){
+	return ids
+		.map(id => id.toLowerCase().trim())
+		.some(id => id === x.toLowerCase().trim());
```
[the remark this answers is not in the document]
Comment (busticated) [reply src/cmd/product.js]: ✅

**Pull request merged**